### ISSUE TITLE
Output comparison with DOMPurify

### ISSUE THREAD

**Issue by doerwalter** [matthiask/html-sanitizer#37]
I've tried the examples that are listed on the [DOMPurify](https://github.com/cure53/DOMPurify) homepage and got this:

| Original HTML                                               | DomPurify output                                      | html_sanitzer output                                   |
| ----------------------------------------------------------- | ----------------------------------------------------- | ------------------------------------------------------ |
| `<img src=x onerror=alert(1)//>`                            | `<img src="x">`                                         | empty string                                         |
| `<svg><g/onload=alert(2)//<p>`                              | `<svg><g></g></svg>`                                    | empty string                                         |
| `<p>abc<iframe//src=jAva&Tab;script:alert(3)>def</p>`       | `<p>abc</p>`                                            | `<p>abcdef</p>`                                      |
| `<math><mi//xlink:href="data:x,<script>alert(4)</script>">` | `<math><mi></mi></math>`                                | `alert(4)"&gt;`                                      |
| `<TABLE><tr><td>HELLO</tr></TABL>`                          | `<table><tbody><tr><td>HELLO</td></tr></tbody></table>` | `HELLO`                                              |
| `<UL><li><A HREF=//google.com>click</UL>`                   | `<ul><li><a href="//google.com">click</a></li></ul>`    | `<ul><li><a href="//google.com">click</a></li></ul>` |

I guess the first two examples are the result of whitelisted elements. The last example does indeed produce the same output for both sanitizer, but the others differ.

Are there any plans to get the outputs to match?

The output for the MathML example looks especially wrong.

**Comment by matthiask** [comment]
Hi

Thanks for the testing! I didn't know DOMPurify but it looks interesting.

As you write the first and second example result in an empty string due to the fact that those elements aren't allowlisted. The same is true for the `<table>` example. 

The iframe example is probably due to my usage of lxml's HTML cleaner, https://lxml.de/apidoc/lxml.html.clean.html, or maybe beautifulsoup; it probably parses the iframe as a self closing tag while DOMPurify treats `def` as the content of the `<iframe>` tag. Not sure what's going on there. I'm not sure which is "better" in this case. In both cases the output is safe and I think that's more important when the HTML is extremely broken.

The MathML example is definitely interesting, yes. I have passed the string into beautifulsoup and here's the result:

```
>>> bs4.BeautifulSoup('''<math><mi//xlink:href="data:x,<script>alert(4)</script>">''')
<html><body><math><mi>alert(4)"&gt;</mi></math></body></html>
```

So the difference results already in the first step, which is (somewhat) outside of the control of html-sanitizer.

Long story short: The reason for the differences are:

- The default configuration of html-sanitizer is more restrictive compared to DOMPurify, 
- The HTML parsing libraries used by html-sanitizer interpret very broken HTML differently.

In both cases the output of DOMPurify doesn't seem to be obviously superior. I'm very open to changing html-sanitizer but you'd also have to show the advantages of doing so for html-sanitizer itself, not just for the interoperability with DOMPurify.

I hope that answers your question.

**Comment by doerwalter** [comment]
I'm not familiar with DOMPurify's source code, but a quick look seems to reveal that DOMPurify uses the browsers `DOMParser`: https://github.com/cure53/DOMPurify/blob/main/src/purify.js#L870

```js
doc = new DOMParser().parseFromString(dirtyPayload, PARSER_MEDIA_TYPE);
```

So DOMPurify's output could be interpreted as being closer to the "official" way of how to parse broken HTML.

But as you're using lxml/Beautifulsoup for the basic parsing, there's probably not much that can be done here.

Thanks for the quick reaction!
